src/traffic_server/traffic_server.cc

```diff
@@ -223,9 +223,19 @@ struct AutoStopCont : public Continuation {
   mainEvent(int /* event */, Event * /* e */)
   {
     APIHook *hook = lifecycle_hooks->get(TS_LIFECYCLE_SHUTDOWN_HOOK);
-    while (hook) {
-      hook->invoke(TS_EVENT_LIFECYCLE_SHUTDOWN, nullptr);
-      hook = hook->next();
+    if (hook != nullptr) {
+      if (hook->m_cont->mutex != nullptr) {
+        MUTEX_TAKE_LOCK(hook->m_cont->mutex, this_ethread());
```

Comment:
Comments in I_Lock.h indicate MUTEX_TAKE_LOCK is deprecated.  In any case, without a corresponding MUTEX_UNTAKE_LOCK invocation, the lock is never released.